pubsub/azure/eventhubs/eventhubs.go

```diff
@@ -15,709 +15,618 @@ package eventhubs
+
+	// Process all partition clients as they come in
+	go func() {
+		for {
+			// This will block until a new partition client is available
+			// It returns nil if processor.Run terminates or if the context is canceled
+			partitionClient := processor.NextPartitionClient(subscribeCtx)
+			if partitionClient == nil {
+				return
+			}
+			aeh.logger.Debugf("Received client for partition %s", partitionClient.PartitionID())
+
+			// Once we get a partition client, process the events in a separate goroutine
+			go func() {
+				processErr := aeh.processEvents(subscribeCtx, req.Topic, partitionClient, eventHandler)
+				// Do not log context.Canceled which happens at shutdown
+				if processErr != nil && !errors.Is(processErr, context.Canceled) {
+					aeh.logger.Errorf("Error processing events from partition client: %v", processErr)
+				}
+			}()
+		}
+	}()
+
+	// Start the processor
+	go func() {
+		// This is a blocking call that runs until the context is canceled
+		err = processor.Run(subscribeCtx)
+		// Do not log context.Canceled which happens at shutdown
+		if err != nil && !errors.Is(err, context.Canceled) {
+			aeh.logger.Errorf("Error from event processor: %v", err)
+		}
+	}()
 
-	if m.ConsumerGroup == "" {
-		return errors.New(missingConsumerIDErrorMsg)
-	}
 	return nil
 }
 
-func (aeh *AzureEventHubs) getStoragePrefixString(topic string) string {
-	// empty string in the end of slice to have a suffix "-".
-	return strings.Join([]string{"dapr", topic, aeh.metadata.ConsumerGroup, ""}, "-")
-}
+func (aeh *AzureEventHubs) processEvents(subscribeCtx context.Context, topic string, partitionClient *azeventhubs.ProcessorPartitionClient, eventHandler func(e *azeventhubs.ReceivedEventData) error) error {
+	// At the end of the method we need to do some cleanup and close the partition client
+	defer func() {
+		closeCtx, closeCancel := context.WithTimeout(context.Background(), resourceGetTimeout)
+		defer closeCancel()
+		closeErr := partitionClient.Close(closeCtx)
+		if closeErr != nil {
+			aeh.logger.Errorf("Error while closing partition client: %v", closeErr)
+		}
+	}()
 
-// Init connects to Azure Event Hubs.
-func (aeh *AzureEventHubs) Init(metadata pubsub.Metadata) error {
-	m, parseErr := parseEventHubsMetadata(metadata)
-	if parseErr != nil {
-		return parseErr
-	}
+	// Loop to receive messages
+	var (
+		ctx    context.Context
+		cancel context.CancelFunc
+		events []*azeventhubs.ReceivedEventData
+		err    error
+	)
+	for {
+		// TODO: Support setting a batch size
+		const batchSize = 1
+		ctx, cancel = context.WithCancel(subscribeCtx)
+		events, err = partitionClient.ReceiveEvents(ctx, batchSize, nil)
+		cancel()
 
-	aeh.metadata = m
-	aeh.eventProcessors = map[string]*eph.EventProcessorHost{}
-	aeh.hubClients = map[string]*eventhub.Hub{}
+		// A DeadlineExceeded error means that the context timed out before we received the full batch of messages, and that's fine
```

Comment:
where is the timeout setting for this particular context `cix`? we are only initialising it with a cancel right?

Reply:
Good catch